### PR TITLE
Add Optimistic Cart behavior

### PR DESCRIPTION
**Pull request by igorbrasileiro**
## What's the purpose of this pull request?
Change the old orderForm provider to use the order-manager and order-items package. These booth packages were extracted from the Checkout IO Apps ([order-items](https://github.com/vtex-apps/order-items) and [order-manager](https://github.com/vtex-apps/order-manager))

## How it works? 
Use both packages.

## How to test it?
Use this version 0.279.0-alpha.7 of `gatsby-theme-store` and `store-ui`

### CODE REVIEW

File: packages/gatsby-theme-store/package.json [hunk 1]
```diff
@@ -33,10 +33,12 @@
...
     "@theme-ui/match-media": "^0.3.1",
-    "@vtex/gatsby-plugin-graphql": "^0.278.0",
-    "@vtex/gatsby-plugin-i18n": "^0.278.0",
-    "@vtex/gatsby-plugin-theme-ui": "^0.278.0",
-    "@vtex/store-ui": "^0.278.0",
+    "@vtex/gatsby-plugin-graphql": "^0.279.0-alpha.1",
+    "@vtex/gatsby-plugin-i18n": "^0.279.0-alpha.1",
+    "@vtex/gatsby-plugin-theme-ui": "^0.279.0-alpha.1",
+    "@vtex/order-items": "^0.0.1-8",
+    "@vtex/order-manager": "^0.0.3-17",
```
Comment (igorbrasileiro) [inline packages/gatsby-theme-store/package.json:40]: Use the right version of these packages

File: packages/gatsby-theme-store/src/components/Minicart/Drawer/Quantity.tsx [hunk 2]
```diff
@@ -29,18 +29,18 @@ const QuantityWrapper: FC<WrapperProps> = ({
...
-  const onChange = async (quantity: number) => {
+  const onChange = (quantity: number) => {
     if (quantity === item.quantity) {
       return
     }
 
     try {
       setIsLoading(true)
-      await updateItems(quantity)
+      updateItems(quantity)
```
Comment (klzns) [inline packages/gatsby-theme-store/src/components/Minicart/Drawer/Quantity.tsx:43]: Can we have that fixed too?

**Review by klzns — APPROVED** [review]

File: packages/gatsby-theme-store/src/components/Minicart/Drawer/Quantity.tsx (hunk 2 above)
Comment (klzns) [reply packages/gatsby-theme-store/src/components/Minicart/Drawer/Quantity.tsx:43]: Ops, not here, in the updateQuantity(0)

Comment (igorbrasileiro) [reply packages/gatsby-theme-store/src/components/Minicart/Drawer/Quantity.tsx:43]: Sure

**Review by klzns — APPROVED** [review]

File: packages/gatsby-theme-store/src/components/Minicart/Drawer/useRemoveItem.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { useOrderItems } from '../../../sdk/orderForm/Provider'
+import { useItem } from './useItem'
+
+export const useRemoveItem = (index: number) => {
+  const { removeItem } = useOrderItems()
+  const item = useItem(index)
+
+  return () => removeItem({ id: `${item.id}`, seller: item.seller, index })
```
Comment (emersonlaurentino) [inline packages/gatsby-theme-store/src/components/Minicart/Drawer/useRemoveItem.ts:8]: Why this?
```suggestion
  return () => removeItem({ id: item.id, seller: item.seller, index })
```

File: packages/gatsby-theme-store/src/components/Minicart/Drawer/useUpdateItem.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { useOrderItems } from '../../../sdk/orderForm/Provider'
+import { useItem } from './useItem'
+
+export const useUpdateItem = (index: number) => {
+  const { updateQuantity } = useOrderItems()
+  const item = useItem(index)
+
+  return (quantity: number) =>
+    updateQuantity({ id: `${item.id}`, quantity, seller: item.seller, index })
```
Comment (emersonlaurentino) [inline packages/gatsby-theme-store/src/components/Minicart/Drawer/useUpdateItem.ts:9]: same
```suggestion
    updateQuantity({ id: item.id, quantity, seller: item.seller, index })
```

File: packages/gatsby-theme-store/src/sdk/orderForm/useLogger.tsx [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { useCallback } from 'react'
+
+export const useLogger = () => {
+  const log = useCallback(() => {}, [])
```
Comment (emersonlaurentino) [inline packages/gatsby-theme-store/src/sdk/orderForm/useLogger.tsx:4]: what is it? 🤔

File: packages/gatsby-theme-store/src/sdk/orderForm/useToast.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// TODO remove this fake useToast
+
+export const useToast = () => {
```
Comment (emersonlaurentino) [inline packages/gatsby-theme-store/src/sdk/orderForm/useToast.ts:3]: this does not should be removed?

Comment (igorbrasileiro) [reply packages/gatsby-theme-store/src/sdk/orderForm/useToast.ts:3]: Nope, this function is required, for now, that we don't have a toast feature we need fake it.

File: packages/gatsby-theme-store/src/sdk/orderForm/useLogger.tsx (hunk 5 above)
Comment (igorbrasileiro) [reply packages/gatsby-theme-store/src/sdk/orderForm/useLogger.tsx:4]: Nope, this function is required, for now, that we don't have a logger we need to fake it for the package.

File: packages/gatsby-theme-store/src/components/Minicart/Drawer/useRemoveItem.ts (hunk 3 above)
Comment (igorbrasileiro) [reply packages/gatsby-theme-store/src/components/Minicart/Drawer/useRemoveItem.ts:8]: id has type of number

File: packages/gatsby-theme-store/src/components/Minicart/Drawer/useUpdateItem.ts (hunk 4 above)
Comment (igorbrasileiro) [reply packages/gatsby-theme-store/src/components/Minicart/Drawer/useUpdateItem.ts:9]: Also here

**Review by emersonlaurentino — APPROVED** [review]

File: packages/gatsby-theme-store/src/sdk/orderForm/Provider.tsx [hunk 7]
```diff
@@ -1,94 +1,75 @@
...
+
+import { useClearOrderFormMessages } from './useClearOrderFormMessage'
+import { useToast } from './useToast'
+import { useLogger } from './useLogger'
+import { useGetOrderForm } from './useGetOrderForm'
+import { useMutateAddItems } from './useMutateAddItems'
+import { useMutateUpdateQuantity } from './useMutateUpdateQuantity'
+import type { OrderFormFragment_OrderFormFragment as OrderForm } from './controller/__generated__/OrderFormFragment_orderForm.graphql'
+
+const defaultOrderForm: OrderForm = {
```
Comment (tlgimenes) [inline packages/gatsby-theme-store/src/sdk/orderForm/Provider.tsx:20]: this could harm performance. We should measure this and see the JSON.parse approach

Comment (igorbrasileiro) [reply packages/gatsby-theme-store/src/sdk/orderForm/Provider.tsx:20]: Do you have a suggestion on how to fix it?

Comment (klzns) [reply packages/gatsby-theme-store/src/sdk/orderForm/Provider.tsx:20]: I think he means this, right?
https://itnext.io/can-json-parse-be-performance-improvement-ba1069951839

Comment (igorbrasileiro) [reply packages/gatsby-theme-store/src/sdk/orderForm/Provider.tsx:20]: But it only enhances performance if use spread operator multiple times, right?

**Pull request merged**